File: bundler-compiler/src/main/java/com/hendraanggrian/bundler/compiler/Names.java
package com.hendraanggrian.bundler.compiler;

import android.support.annotation.NonNull;

import com.google.auto.common.MoreElements;
import com.google.common.collect.Lists;
import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.TypeName;

import java.util.Collections;
import java.util.List;

import javax.lang.model.element.TypeElement;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
final class Names {

    // internal
    static final ClassName CLASS_BUNDLE_BINDING;
    static final ClassName CLASS_BUNDLER_UTILS;
    // Android
    static final ClassName CLASS_BUNDLE;
    static final ClassName CLASS_PARCELABLE;
    static final ClassName CLASS_SPARSE_ARRAY;
    // Parceler
    static final ClassName CLASS_PARCEL;
    static final ClassName CLASS_PARCELS;

    static final String TARGET = "target";
    static final String SOURCE = "source";
    static final String ARGS = "args";

    static {
        final String bundler = "com.hendraanggrian.bundler";
        CLASS_BUNDLE_BINDING = ClassName.get(bundler, "BundleBinding");
        CLASS_BUNDLER_UTILS = ClassName.get(bundler, "BundlerUtils");

        final String android = "android";
        final String androidOs = android + ".os";
        final String androidUtil = android + ".util";
        CLASS_BUNDLE = ClassName.get(androidOs, "Bundle");
        CLASS_PARCELABLE = ClassName.get(androidOs, "Parcelable");
        CLASS_SPARSE_ARRAY = ClassName.get(androidUtil, "SparseArray");

        final String parceler = "org.parceler";
        CLASS_PARCEL = ClassName.get(parceler, "Parcel");
        CLASS_PARCELS = ClassName.get(parceler, "Parcels");
    }

    @NonNull
    static String guessGeneratedName(@NonNull TypeElement typeElement, @NonNull String suffix) {
        List<String> enclosings = Lists.newArrayList(typeElement.getSimpleName().toString());
        while (typeElement.getNestingKind().isNested()) {
            typeElement = MoreElements.asType(typeElement.getEnclosingElement());
            enclosings.add(typeElement.getSimpleName().toString());
        }
        Collections.reverse(enclosings);
        String typeName = enclosings.get(0);
        for (int i = 1; i < enclosings.size(); i++)
            typeName += '$' + enclosings.get(i);
        return typeName + suffix;
    }

    @NonNull
    static TypeName safeUnbox(@NonNull TypeName input) {
        if (input.isBoxedPrimitive())
            return input.unbox();
        return input;
    }
}

File: bundler-compiler/src/main/kotlin/com/hendraanggrian/bundler/compiler/Names.kt
package com.hendraanggrian.bundler.compiler

import com.google.auto.common.MoreElements
import com.squareup.javapoet.ClassName.get
import java.util.*
import javax.lang.model.element.TypeElement

internal const val TARGET = "target"
internal const val SOURCE = "source"
internal const val ARGS = "args"
internal const val KEY = "key"
internal const val VALUE = "value"
internal const val DEFAULT_VALUE = "defaultValue"
internal const val GENERIC = "T"

internal val TYPE_BUNDLE_BINDING = get("com.hendraanggrian.bundler.internal", "BundleBinding")!!
internal val TYPE_BUNDLER_UTILS = get("com.hendraanggrian.bundler", "BundlerUtils")!!

internal val TYPE_PARCEL = get("org.parceler", "Parcel")!!
internal val TYPE_PARCELS = get("org.parceler", "Parcels")!!

internal fun TypeElement.getMeasuredName(suffix: String): String {
    val enclosings = mutableListOf(simpleName.toString())
    var typeElement = this
    while (typeElement.nestingKind.isNested) {
        typeElement = MoreElements.asType(typeElement.enclosingElement)
        enclosings.add(typeElement.simpleName.toString())
    }
    Collections.reverse(enclosings)
    var typeName = enclosings[0]
    for (i in 1 until enclosings.size) typeName += "$${enclosings[i]}"
    return "$typeName$suffix"
}